### DISCUSSION TITLE
what is the correct way to obtain an ast from a value ?

### DISCUSSION THREAD

**Discussion by yassinm** [cue-lang/cue#1662] [Q&A] [answered]
Hi

I am trying to recover where a particular `cue.Value` came from in the current api. As in which dir, file, package. Is that possible ?

**Comment by verdverm** [comment]
Somewhat, there are cue.Value Source and Syntax functions, and the nodes will have a Pos function.

The difficulty is that a value may be defined in multiple locations, possibly across files or directories

**Reply by yassinm** [reply]
I imagine the pos could also get complicated if the value came from a list comprehension ? I will try that and report back.

**Comment by myitcv** [comment, marked as answer]
@yassinm as @verdverm suggested, you can use `cue.Value.Source()` to get an `cue/ast.Node` that then gives you the position information:

```
go mod tidy
go run main.go

-- go.mod --
module mod.com

go 1.18

require cuelang.org/go v0.4.3

require (
	github.com/cockroachdb/apd/v2 v2.0.1 // indirect
	github.com/emicklei/proto v1.6.15 // indirect
	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b // indirect
	github.com/google/uuid v1.2.0 // indirect
	github.com/mpvl/unique v0.0.0-20150818121801-cbe035fff7de // indirect
	github.com/pkg/errors v0.8.1 // indirect
	github.com/protocolbuffers/txtpbfmt v0.0.0-20201118171849-f6a6b3f636fc // indirect
	golang.org/x/net v0.0.0-20200226121028-0de0cce0169b // indirect
	golang.org/x/text v0.3.7 // indirect
	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
)
-- main.go --
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

func main() {
	ctx := cuecontext.New()
	bps := load.Instances([]string{"."}, nil)
	v := ctx.BuildInstance(bps[0])

	x := v.LookupPath(cue.ParsePath("x"))
	y := v.LookupPath(cue.ParsePath("y"))

	xsyn := x.Source()
	ysyn := y.Source()

	fmt.Printf("x: %v\n", xsyn.Pos())
	fmt.Printf("y: %v\n", ysyn.Pos())
}
-- x.cue --
package z

x: 5

y: int
y: 5 & x
```

This gives:

```
x: $WORK/x.cue:3:1
y: $WORK/x.cue:6:4
```

As you can see the only wrinkle here is that we might consider the first of the `y` declarations to be its position. Here, the position returned for `y` is the second declaration.  

Can you provide a bit more detail about what you are ultimately trying to achieve here?

Note there is also [`cue.Value.Expr()`](https://pkg.go.dev/cuelang.org/go/cue#Value.Expr) that breaks a `cue.Value` down into its component parts, in case this is what you are after.

**Reply by yassinm** [reply]
@myitcv I have few use cases for this.

1) error reporting so i can tell the user where the actual error came from (if i can). 
2) in a visual editor telling the user where a value came from.
3) figuring where to insert delete/elements in a file if a tool wants to delete/augment the cue "graph" [1]
4) merging/reconstructing modified cue files from a "cache" while preserving file locations (for easier git diffs) [2]

[1] user wants to pass a flag to a cli tool to remove/override a value in a cue file. As in `my-cli --set-value <key>:<value>`

[2] the tool i have is a cli tool that wants to read cue files and store part of the the resulting (unified) value in a db as json (for example). The simplest option would be to store the actual files on the user disk in the db as opposed to using json which is not optimal. Also the local fs might contain more than what the user wants to store in the db locally :-( The question then is how much metadata do i need to store in the resulting json tree so that i can reconstruct what was on disk later and the user can do a git diff locally ? Again there is a quick and dirty solution albeit non optimal.

I imagine 1,2 would be the easiest and i will try using the APIs you provided which should cover my use case and my immediate need for now. 3,4 are maybe more involved scenarios.  As you mentioned its not super straight forward. I might be forced to go down to the parser level for these and avoid dealing with unified values but i am not there yet ! I will first make sure i can use your suggestions and report my findings.

Thank you @verdverm  and @myitcv for taking the time to answer this. Really appreciated !

**Reply by yassinm** [reply]
I had a shower thought about 4. and i think i have a simplification now. If it was not clear already here is the use case.
* user edits cue files locally
* user uses the cli to unify all values (or part of the graph) and stores the resulting json in a db
* time goes on and user edits their files locally ... or someone else does the same (or different) operation on their fs and stores a different graph in the db
* user wants to know what is the difference between what they have locally and what is stored in the db

At first i was thinking to reconstruct what is stored locally starting from what is stored remotely then do a textual diff. I am thinking a simpler approach would be to first do the same (unification) operation locally from the local fs then do a textual[1] comparison with what is stored remotely. This way i only have to show the user where what they have locally differs from what is stored remotely. As such its the same as use case 2. I have no need to store metadata or cue files remotely !

[1] textual at first but might get more involved later

**Reply by verdverm** [reply]
@yassinm You might find `hof dm` and the Diff helper interesting. This does a structural diff on CUE values, using the GoAPI

- https://docs.hofstadter.io/first-example/adding-a-datamodel/checkpointing/
- https://github.com/hofstadter-io/hof/blob/_dev/lib/structural/diff.go

The basic idea is to store a cue.Value's complete history in local files, like: https://github.com/hofstadter-io/hof/tree/_dev/.hof/dm/MyModels and then diff them as needed

The reason to store history checkpoints is so that you can see the entire history within a single execution, without having to navigate git. This is what makes it possible for hof to generate all migrations needed for a database over time.

You should be able to reuse the Diff function for any purpose though!

```cue
// 20211230215755 -> dirty-20220426155614
MyModels: {
        Models: {
                User: {
                        Fields: {
                                "+": {  // this is an addition, there is also a "-" field
                                        username: {
                                                Name:     "username"
                                                Type:     "string"
                                                Length:   64
                                                Unique:   false
                                                Nullable: false
                                                Validation: {
                                                        Max: 64
                                                }
                                        }
                                }
                        }
// ...
```

**Reply by myitcv** [reply]
> user wants to pass a flag to a cli tool to remove/override a value in a cue file.

What are you trying to do in this situation? I would caution against providing a mechanism outside of CUE to remove/override a value. Because that very much goes against the ethos of no inheritance. Is there, instead, a different way of solving the example you have in mind here?

For diff support please also see https://github.com/cue-lang/cue/issues/8

**Reply by yassinm** [reply]
@myitcv the idea is to avoid asking a user to open an editor and edit a file in place. Think more like a sed tool for a cue file more than anything.

If this was a yaml file as you said this would be grep/search to remove the value because values are all concrete. The problem is in cue  there is no easy way to find where a particular value came from. I leave it up to you if you think this is a common case or not but to me it seems like a common case.

**Reply by myitcv** [reply]
> The problem is in cue there is no easy way to find where a particular value came from.

Can you provide some specific examples of where and how this is challenging?

The reason for me asking is that tooling, specifically the [LSP](https://github.com/cue-lang/cue/issues/142), will help answer these sorts of questions/problems. CUE actually makes finding the source of values possible, whereas in the case of inheritance/overlays etc the problem is not generally solvable. 

The issue therefore with allowing the user to provide overrides/remove values is that you introduce a mechanism _outside of CUE_ which means your users need to look elsewhere to find the answer of "why is this field this value?"

Just to check, have you seen that CUE supports injection of values?

```
$ cue help injection
Many of the cue commands allow injecting values or
selecting files from the command line using the --inject/-t flag.
...
```

**Reply by yassinm** [reply]
> Can you provide some specific examples of where and how this is challenging?

I think if if i continue using an in place edit mentality (ala sed) it will become challenging. For example if the value came from a list comprehension then in place editing (meaning replace the value as in sed) is challenging.

> The reason for me asking is that tooling, specifically the LSP, will help answer these sorts of questions/problems.

Originally i was leaning towards a sed tool and not go for a full blown editor with LSP support.

> Just to check, have you seen that CUE supports injection of values?

I have not seen that until now so thank you ! I guess what i am hearing is to rethink the problem in terms of injecting/weaving in values instead of search/replace.

Really appreciate all your input :-)

**Reply by myitcv** [reply]
> I guess what i am hearing is to rethink the problem in terms of injecting/weaving in values instead of search/replace.

Exactly. But if you run into issues, or find the pattern awkward, please say: happy to help work through examples in case it points to improvements that can be made on the CUE side.

**Comment by mpvl** [comment]
A "composed" value may not have a source. One way to get at the sources is to "split" a value until it has a source or can no longer be split using `value.Expr()`.

We may consider using a separate position API, but we probably should do some more work on errors first to see how such an API could look like. For instance, it might also be useful to get access to positions that are not directly related to a value: if a value comes from a definition, for instance, it may also be worth knowing the point at which that definition was included, which may be higher up in the tree. All this information is accessible from a `Value`, it is just not exposed in the API at the moment.
